runtime/sema/type.go

```diff
@@ -3779,6 +3827,137 @@ func (t *CompositeType) GetNestedTypes() *StringTypeOrderedMap {
 	return t.NestedTypes
 }
 
+func (t *CompositeType) isTypeIndexableType() bool {
+	// resources and structs only can be indexed for attachments
+	return t.Kind.SupportsAttachments()
+}
+
+func (t *CompositeType) TypeIndexingElementType(indexingType Type) Type {
+	return &OptionalType{
+		&ReferenceType{
+			Type: indexingType,
+		},
+	}
+}
+
+func (t *CompositeType) IsValidIndexingType(ty Type) bool {
+	attachmentType, isComposite := ty.(*CompositeType)
+	return isComposite &&
+		IsSubType(t, attachmentType.baseType) &&
+		attachmentType.IsResourceType() == t.IsResourceType()
+}
+
+const compositeForEachAttachmentFunctionName = "forEachAttachment"
+
+const compositeForEachAttachmentFunctionDocString = `
+Iterates over the attachments present on the receiver, applying the function argument to each.
+The order of iteration is undefined. If a type argument is provided, only attachments that conform
+to the specified interface are iterated on, the others are filtered out. 
+`
+
+func CompositeForEachAttachmentFunctionType(t CompositeKindedType) *FunctionType {
+	attachmentSuperType := AnyStructAttachmentType
+	if t.IsResourceType() {
+		attachmentSuperType = AnyResourceAttachmentType
+	}
+
+	return &FunctionType{
+		Parameters: []*Parameter{
+			{
+				Label:      ArgumentLabelNotRequired,
+				Identifier: "f",
+				TypeAnnotation: NewTypeAnnotation(
+					&FunctionType{
+						Parameters: []*Parameter{
+							{
+								TypeAnnotation: NewTypeAnnotation(
+									&ReferenceType{
+										Type: attachmentSuperType,
+									},
+								),
+							},
+						},
+						ReturnTypeAnnotation: NewTypeAnnotation(VoidType),
+					},
+				),
+			},
+		},
+		ReturnTypeAnnotation: NewTypeAnnotation(VoidType),
+	}
+}
+
+const attachmentGetFieldFunctionName = "getField"
+
+const attachmentGetFieldFunctionDocString = `
+Returns a reference to the field with the specified name and type on the receiver. If a field with the
+specified name does not exist on the receiver, or if the field exists but does not exactly match the provided type, 
+returns nil.
+`
+
+func AttachmentGetFieldFunctionType() *FunctionType {
+	typeParameter := &TypeParameter{
+		Name:      "T",
+		TypeBound: AnyType,
+	}
+
+	return &FunctionType{
+		TypeParameters: []*TypeParameter{
+			typeParameter,
+		},
+		Parameters: []*Parameter{
+			{
+				Label:          ArgumentLabelNotRequired,
+				Identifier:     "name",
+				TypeAnnotation: NewTypeAnnotation(StringType),
+			},
+		},
+		ReturnTypeAnnotation: NewTypeAnnotation(
+			&OptionalType{
+				Type: &ReferenceType{
+					Type: &GenericType{
+						TypeParameter: typeParameter,
+					},
+				},
+			},
+		),
+	}
+}
+
+const attachmentGetFunctionFunctionName = "getFunction"
```

Comment:
Might want this to be an `invokeFunction` instead, which would take the name and a variadic number of arguments depending on the type parameter, and return an optional return type. This would prevent issues with bound resource functions.